src/cli/services/linter/utils/deduplicateResults.ts

```diff
@@ -1,26 +1,28 @@
-import { IRange } from '@stoplight/types/dist';
+import { resultsComparer } from '../../../../formatters/utils/sortResults';
 import { IRuleResult } from '../../../../types';
 
-type Dictionary<T, K extends PropertyKey> = { [key in K]: T };
+export const deduplicateResults = (results: IRuleResult[]): IRuleResult[] => {
+  const filtered: IRuleResult[] = [];
 
-const ARTIFICIAL_ROOT = Symbol('root');
+  const howMany = results.length;
 
-const serializeRange = ({ start, end }: IRange) => `${start.line}:${start.character}:${end.line}:${end.character}`;
-const getIdentifier = (result: IRuleResult) => `${result.path.join('/')}${result.code}${serializeRange(result.range)}`;
+  if (howMany < 2) {
+    return [...results];
+  }
 
-export const deduplicateResults = (results: IRuleResult[]) => {
-  const seen: Dictionary<Dictionary<string, string>, symbol> = {};
+  const sorted = results.sort(resultsComparer);
 
-  return results.filter(result => {
-    const source = result.source === void 0 ? ARTIFICIAL_ROOT : result.source;
-    const identifier = getIdentifier(result);
-    if (!(source in seen)) {
-      seen[source] = {};
-    } else if (identifier in seen[source]) {
-      return false;
+  for (let i = 0; i < howMany - 1; i++) {
+    if (resultsComparer(sorted[i], sorted[i + 1]) === 0) {
+      continue;
     }
 
-    seen[source][identifier] = true;
-    return true;
-  });
+    filtered.push(sorted[i]);
+  }
+
+  if (resultsComparer(sorted[howMany - 1], filtered[filtered.length - 1]) !== 0) {
```

Comment:
Technically there is a possibility of out of bounds access here (in case of `filtered` array).
Let's avoid it.
```suggestion
  if (filtered.length > 0 && resultsComparer(sorted[howMany - 1], filtered[filtered.length - 1]) !== 0) {
```

Reply:
Hmmm. It's possible I'm lacking a test case here. Let me see how I can put that under the light.

Reply:
@P0lip There was a bug. And thanks to you I've found it. Added a test case and changed the algo to make it backward looking (rather than forward looking). It's also come with a slight benefit of making the whole thing easier to understand (and hopefully review).